### ISSUE TITLE
Make StatusQ a QML module

### ISSUE THREAD

**Issue by igor-sirotin** [status-im/status-desktop#9863]
# TLDR;

All we have in StatusQ are QML components. Implemented in QML or C++.
This could be compiled into a [QML module](https://doc.qt.io/qt-5/qtqml-modules-topic.html), which would be loaded dynamically by QQmlEngine when `import StatusQ` is found.

Good things have been done in https://github.com/status-im/status-desktop/issues/9125, but it seems to me that we can do better 💪 

![image](https://user-images.githubusercontent.com/25482501/224862117-b2179306-ff68-4d73-bc36-a812f423330f.png)

# What's the problem?

`StatusQ` doesn't provide a "binary product".
It's simply a repository with C++ and QML components. 
It's also has `SortFilterProxyModel` as a submodule and require those to be linked after all.

Because of this, we have to compile StatusQ (C++ components, QML, dependeices like SortFilterProxyModel) as part of something else.

E.g., this is how we link StatusQ to status-desktop:
1. Compile StatusQ C++ components as part of `DOtherSide`
2. Compile `SortFilterProxyModel` as part of `DOtherSide`
3. Generate a RCC with QML files, dynamically load it
4. Call StatusQ `registerTypes`

And we have to do this manually for EACH usage of StatusQ in:
- DOtherSide (this one registers things in the Nim app)
- sandbox
- storybook
- StatusQ sanity checker

And it's done in a different way every time.

## Dude, just give some images!

![image](https://user-images.githubusercontent.com/25482501/224862084-fe85dac3-3547-4321-b3f5-9ef9df9e2c24.png)

# Registering QML types

I recently added https://github.com/status-im/qzxing as dependency for QR code scanning (https://github.com/status-im/status-desktop/pull/9464). It was a nightmare adding it.

I thought that `SortFilterProxyModel` would be a great guide for me on how to link such dependency.
What I didn't know was:
- `SortFilterProxyModel` is internally built as `OBJECT` library
- `SortFilterProxyModel` is internally using `Q_COREAPP_STARTUP_FUNCTION` macro to register types

Both of this clauses are false for `QZXing`.
1. It’s internally built as `STATIC` library
1. You should manually call `QZXing::registerQMLTypes` for the types to be found in QML

So, comparing to `SortFilterProxyModel`, here I need to:
- link `QZXing` to each usage
- call `QZXing::registerQMLTypes()`

# How does QML module fix this?

Each QML module can override `initializeEngine` function that’s being called on module loading.
This function is intended to register all cpp-implemented QML elements with appropriate library name.

🎉  This way `QZXing` and `SortFilterProxyModel` could be pure dependencies of `StatusQ`!
And we would call to `QZXing::registerQMLTypes` on module loading.

`StatusQ` knows nothing about `DOtherSide`. Nor should `DOtherSide` know anything about `StatusQ`.
We should try hard to make those libraries independent. 
And simply connect required libraries directly to:
- status desktop
- sandbox
- storybook
- StatusQ sanity checker

**Comment by caybro** [comment]
Fully agree :) Just for reference, this is exactly what I've been doing for the new StatusQ theme/style engine: https://github.com/caybro/StatusQ2/blob/master/imports/StatusQ/Core/Theme/plugin.cpp 

Such a plugin "registers" itself and then on the QML side it's just sufficient to:
```qml
import StatusQ.Core.Theme 1.0
...
```

No need to link the app binary to the plugin DLL at all

**Comment by caybro** [comment]
Also worth noting, that no QML types have to be manually exported at all, it happens via the Qt macros (`QML_ELEMENT` and `QML_SINGLETON` here) and some qmake magic; example https://github.com/caybro/StatusQ2/blob/master/imports/StatusQ/Core/Theme/theme.h:

```cpp
#include <QmlTypeAndRevisionsRegistration>
...
class Theme : public QObject
{
  Q_OBJECT
  QML_ELEMENT
  QML_SINGLETON
  Q_DISABLE_COPY_MOVE(Theme)
...
```

**Comment by osmaczko** [comment]
Go for it :+1: 

> SortFilterProxyModel is internally using Q_COREAPP_STARTUP_FUNCTION macro to register types

In general yes, but for status-desktop the manual registration is needed anyway: https://github.com/status-im/status-desktop/blob/3f66e41929465cf09d17e01bd12e7f9bc9c0f756/vendor/DOtherSide/lib/src/DOtherSide.cpp#L106
